docs/Migrating-Extensions-v2.0.md

```diff
@@ -0,0 +1,210 @@
+  new features. In the `1.x` branch, `1.8` was never released outside of a "beta" version for this reason: changes in the
+  `protected` API layer would break extensions.
+
+In order to address these concerns, "extensions" in Parsedown v2.0 will become more like "plugins", and with that comes a lot of
+flexability.
+
+ParsedownExtra is a popular extension for Parsedown, and this has been completely re-implemented for 2.0. In order to use
+ParsedownExtra with Parsedown, a user simply needs to write the following:
+
+```php
+$Parsedown = new Parsedown(new ParsedownExtra);
+$actualMarkup = $Parsedown->toHtml($markdown);
+```
+
+Here, ParsedownExtra is *composed* with Parsedown, but does not extend it.
+
+A key feature of *composability* is the ability to compose *multiple* extensions together, for example another
+extension, say, `ParsedownMath` could be composed with `ParsedownExtra` in a user-defined order.
+
+This time using the `::from` method, rather than the convinence constructor provided by `ParsedownExtra`.
+
+```php
+$Parsedown = new Parsedown(ParsedownExtra::from(ParsedownMath::from(new State)));
+```
+
+```php
+$Parsedown = new Parsedown(ParsedownMath::from(ParsedownExtra::from(new State)));
+```
+
+## Introduction to the `State` Object
+Key to Parsedown's new composability for extensions is the `State` object.
+
+This name is a little obtuse, but is importantly accurate.
+
+A `State` object incorporates `Block`s, `Inline`s, some additional render steps, and any custom configuration options that
+the user might want to set. This can **fully** control how a document is parsed and rendered.
+
+In the above code, `ParsedownExtra` and `ParsedownMath` would both be implementing the `StateBearer` interface, which
+essentially means "this class holds onto a particular Parsedown State". A `StateBearer` should be constructable from
+an existing `State` via `::from(StateBearer $StateBearer)`, and reveals the `State` it is holding onto via `->state(): State`.
+
+Implementing the `StateBearer` interface is **strongly encouraged** if implementing an extension, but not necessarily required.
+In the end, you can modify Parsedown's behaviour by producing an appropriate `State` object (which itself is trivially a
+`StateBearer`).
+
+In general, extensions are encouraged to go further still, and split each self-contained piece of functionality out into its own
+`StateBearer`. This will allow your users to cherry-pick specific pieces of functionality and combine it with other
+functionality from different authors as they like. For example, a feature of ParsedownExtra is the ability to define and expand
+"abbreviations". This feature is self-contained, and does not depend on other features (e.g. "footnotes").
+
+A user could import *only* the abbreviations feature from ParsedownExtra by using the following:
+
+```php
+use Erusev\ParsedownExtra\Features\Abbreviations;
+
+$State = Abbreviations::from(new State);
```

Comment:
We have same issue here `new State` will by itself return a error so some info about how to handle it will help

Reply:
You're right, I should also be importing `State` next to the other `use` line :)

Reply:
This should be resolved in bf6cfe8